### PR TITLE
[Merged by Bors] - chore:  `prodAssoc` as an additive/multiplicative or linear equivalence

### PR DESCRIPTION
**Pull request by robin-carlier**
Record multiplicativity/linearity of the equivalence `(E × F) × G ≃ E × (F × G)`. Adapt the construction of `LinearIsometryEquiv.prodAssoc` to use this.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Group/Prod.lean [hunk 1]
```diff
@@ -587,7 +587,24 @@ theorem coe_prodComm : ⇑(prodComm : M × N ≃* N × M) = Prod.swap :=
 theorem coe_prodComm_symm : ⇑(prodComm : M × N ≃* N × M).symm = Prod.swap :=
   rfl
 
-variable {M' N' : Type*} [MulOneClass M'] [MulOneClass N']
+variable {M' : Type*} [MulOneClass M']
+
+/-- The equivalence between `(M × N) × M'` and `M × (N × M')` is multiplicative. -/
+@[to_additive prodAssoc
+      "The equivalence between `(M × N) × M'` and `M × (N × M')` is additive."]
+def prodAssoc : (M × N) × M'≃* M × (N × M') :=
```
Comment (jcommelin) [inline Mathlib/Algebra/Group/Prod.lean:595]:
```suggestion
def prodAssoc : (M × N) × M' ≃* M × (N × M') :=
```

File: Mathlib/Algebra/Group/Prod.lean [hunk 2]
```diff
@@ -587,7 +587,24 @@ theorem coe_prodComm : ⇑(prodComm : M × N ≃* N × M) = Prod.swap :=
...
+  { Equiv.prodAssoc M N M' with map_mul' := fun ⟨_, _⟩ ⟨_, _⟩ => rfl }
+
+@[to_additive (attr := simp) coe_prodAssoc]
+theorem coe_prodAssoc : ⇑(prodAssoc : (M × N) × M'≃* M × (N × M')) = Equiv.prodAssoc M N M' :=
```
Comment (jcommelin) [inline Mathlib/Algebra/Group/Prod.lean:599]:
```suggestion
theorem coe_prodAssoc : ⇑(prodAssoc : (M × N) × M' ≃* M × (N × M')) = Equiv.prodAssoc M N M' :=
```

File: Mathlib/Algebra/Group/Prod.lean [hunk 3]
```diff
@@ -587,7 +587,24 @@ theorem coe_prodComm : ⇑(prodComm : M × N ≃* N × M) = Prod.swap :=
...
+  rfl
+
+@[to_additive (attr := simp) coe_prodAssoc_symm]
+theorem coe_prodAssoc_symm :
+    ⇑(prodAssoc : (M × N) × M'≃* M × (N × M')).symm = (Equiv.prodAssoc M N M').symm :=
```
Comment (jcommelin) [inline Mathlib/Algebra/Group/Prod.lean:604]:
```suggestion
    ⇑(prodAssoc : (M × N) × M' ≃* M × (N × M')).symm = (Equiv.prodAssoc M N M').symm :=
```

File: Mathlib/LinearAlgebra/Prod.lean [hunk 4]
```diff
@@ -657,6 +657,31 @@ theorem snd_comp_prodComm :
...
+
+theorem fst_comp_prodAssoc :
+    (LinearMap.fst R M (M₂ × M₃)).comp (prodAssoc R M M₂ M₃).toLinearMap =
+    (LinearMap.fst R M M₂).comp (LinearMap.fst R (M × M₂) M₃) := by
+  ext <;> simp
+
+theorem snd_comp_prodAssoc :
+    (LinearMap.snd R M (M₂ × M₃)).comp (prodAssoc R M M₂ M₃).toLinearMap =
+    (LinearMap.snd R M M₂).prodMap (LinearMap.id : M₃ →ₗ[R] M₃):= by
+  ext <;> simp
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/Prod.lean:681]: Could you please use `M₁` here as well? (For aesthetic reasons.)

**Review by jcommelin — COMMENTED** [review]
I personally would prefer `M₁ M₂ M₃` over `M N M'` for the mulEquiv. It just looks better.

**Comment by robin-carlier** [comment]
> I personally would prefer `M₁ M₂ M₃` over `M N M'` for the mulEquiv. It just looks better.

Subscripts notations are nowhere used within `Algebra/Group/Prod`, Should I introduce them here just for this case?

**Comment by robin-carlier** [comment]
I ended up going with `M`, `N`and `P` in `Algebra/Group/Prod.lean`, which I feel is a middle ground between a lone primed term and introducing completely new notations in the file.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**